src/repr/binary_codec.ml

```diff
@@ -0,0 +1,467 @@
+    | Static n, `Fixed str_len -> Sizer.static (n + str_len)
+    | _, _ -> (
+        let decode_len = unstage (Len.decode header_typ) in
+        let of_encoding buf (Size.Offset off) =
+          let off, size = decode_len buf off in
+          assert (size >= 0);
+          Size.Offset (off + size)
+        in
+        match size_of_header with
+        | Unknown -> assert false
+        | Static n ->
+            Sizer.dynamic ~of_encoding ~of_value:(fun s -> n + length s)
+        | Dynamic f ->
+            Sizer.dynamic ~of_encoding ~of_value:(fun s ->
+                let s_len = length s in
+                f s_len + s_len))
+end
+
+module String_unboxed = struct
+  let encode _ = stage (fun s k -> k s)
+
+  let decode _ =
+    Mono_container.decode_unboxed (fun x -> x) Bytes.unsafe_to_string
+
+  let sizer n = Mono_container.sizer_unboxed ~length:String.length n
+end
+
+module Bytes_unboxed = struct
+  (* NOTE: makes a copy of [b], since [k] may retain the string it's given *)
+  let encode _ = stage (fun b k -> k (Bytes.to_string b))
+
+  let decode _ =
+    Mono_container.decode_unboxed Bytes.unsafe_of_string (fun x -> x)
+
+  let sizer n = Mono_container.sizer_unboxed ~length:Bytes.length n
+end
+
+module String = struct
+  let encode len =
+    let encode_len = unstage (Len.encode len) in
+    stage (fun s k ->
+        let i = String.length s in
+        encode_len i k;
+        k s)
+
+  let decode len = Mono_container.decode (fun x -> x) Bytes.unsafe_to_string len
+  let sizer n = Mono_container.sizer ~length:String.length n
+end
+
+module Bytes = struct
+  let encode len =
+    let encode_len = unstage (Len.encode len) in
+    stage (fun s k ->
+        let i = Bytes.length s in
+        encode_len i k;
+        unsafe_add_bytes s k)
+
+  let decode len = Mono_container.decode Bytes.unsafe_of_string (fun x -> x) len
+  let sizer len = Mono_container.sizer ~length:Bytes.length len
+end
+
+module Option = struct
+  let encode encode_elt v k =
+    match v with
+    | None -> Char.encode '\000' k
+    | Some x ->
+        Char.encode '\255' k;
+        encode_elt x k
+
+  let decode decode_elt buf ofs =
+    let ofs, c = Char.decode buf ofs in
+    match c with
+    | '\000' -> (ofs, None)
+    | _ ->
+        let ofs, x = decode_elt buf ofs in
+        (ofs, Some x)
+
+  let sizer : type a. a Sizer.t -> a option Sizer.t =
+   fun elt ->
+    let header_size = 1 in
+    match elt with
+    | { of_value = Static 0; _ } ->
+        Sizer.static header_size (* Either '\000' or '\255' *)
+    | { of_value = Static n; _ } ->
+        (* Must add [n] in the [Some _] case, otherwise just header size. *)
+        let of_value = function
+          | None -> header_size
+          | Some _ -> header_size + n
+        in
+        let of_encoding buf (Size.Offset off) =
+          match Stdlib.String.get buf off with
+          | '\000' -> Size.Offset (off + 1)
+          | _ -> Size.Offset (1 + n)
```

Comment:
```suggestion
          | '\000' -> Size.Offset (off + header_size)
          | _ -> Size.Offset (off + header_size + n)
```
I must be wrong but shouldn't we add `off` in the `_` case?

Reply:
You are not wrong! This was a bug. Nice spot :slightly_smiling_face:

Reply:
(Currently adding some better tests for this...)